File: elixir/word-count/word-count.exs
defmodule Words do
  @doc """
  Counts the occurences of each word in the given phrase treating
  each word as case-insensitive.  
  """
  def count(string) do
    words = string |> Text.as_words
    count_words(words)
  end
        
  defp count_words(words) do
    Enum.reduce words, HashDict.new, count_word(&1, &2)
  end
  
  defp count_word(word, dict) do
    HashDict.update dict, word |> String.downcase, 1, &1 + 1
  end
end

defmodule Text do
  @doc """
  Converts a given string into a collection of alphanumeric words.
  """
  def as_words(string) do
    string
      |> String.split(%r{\W}) 
      |> Enum.reject(is_empty? &1)    
  end
  
  defp is_empty?(string), do: String.length(string) == 0
end

File: ruby/word-count/word_count.rb
class Phrase < String
  def word_count
    words.each_with_object(Hash.new(0)) do |word, count|
      filtered_word = word.gsub(/^'|'$/, "")
      count[filtered_word] += 1 
    end
  end

  def words
    downcase.split(/[^'a-zA-Z0-9]+/)
  end
end
